hedera-node/test-clients/src/main/java/com/hedera/services/bdd/junit/support/validators/block/BlockContentsValidator.java

```diff
@@ -0,0 +1,121 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hedera.services.bdd.junit.support.validators.block;
+
+import static com.hedera.services.bdd.junit.hedera.utils.WorkingDirUtils.workingDirFor;
+
+import com.hedera.hapi.block.stream.Block;
+import com.hedera.hapi.block.stream.BlockItem;
+import com.hedera.services.bdd.junit.support.BlockStreamAccess;
+import com.hedera.services.bdd.junit.support.BlockStreamValidator;
+import com.hedera.services.bdd.spec.HapiSpec;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import java.nio.file.Paths;
+import java.util.List;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.junit.jupiter.api.Assertions;
+
+public class BlockContentsValidator implements BlockStreamValidator {
+    private static final Logger logger = LogManager.getLogger(BlockContentsValidator.class);
+
+    public static void main(String[] args) {
+        final var node0Dir = Paths.get("hedera-node/test-clients")
+                .resolve(workingDirFor(0, "hapi"))
+                .toAbsolutePath()
+                .normalize();
+        final var validator = new BlockContentsValidator();
+        final var blocks =
+                BlockStreamAccess.BLOCK_STREAM_ACCESS.readBlocks(node0Dir.resolve("data/block-streams/block-0.0.3"));
+        validator.validateBlocks(blocks);
+    }
+
+    public static final Factory FACTORY = new Factory() {
+        @NonNull
+        @Override
+        public BlockStreamValidator create(@NonNull final HapiSpec spec) {
+            return new BlockContentsValidator();
+        }
+    };
+
+    @Override
+    public void validateBlocks(@NonNull final List<Block> blocks) {
+        for (int i = 0; i < blocks.size(); i++) {
+            try {
+                validate(blocks.get(i), i == 0);
+            } catch (AssertionError err) {
+                logger.error("Error validating block {}", blocks.get(i));
+                throw err;
+            }
+        }
+    }
+
+    private static void validate(Block block, boolean isFirst) {
+        final var blockItems = block.items();
+
+        // A block SHALL start with a `header`.
+        if (!blockItems.getFirst().hasBlockHeader()) {
+            Assertions.fail("Block does not start with a block header");
+        }
+
+        // A block SHALL start with a `header`.
+        if (!blockItems.getFirst().hasBlockHeader()) {
+            Assertions.fail("Block does not start with a block header");
+        }
+
+        // A block SHALL end with a `state_proof`.
+        if (!blockItems.getLast().hasBlockProof()) {
+            Assertions.fail("Block does not end with a block proof");
+        }
+
+        // In general, a `block_header` SHALL be followed by an `round_header`, but for hapiTestRestart we get
```

Comment:
Given this comment, should this validator account for a state change being present?

Reply:
Missed to update the comment, this now applies to the round header check.

Reply:
Updated.